### PR TITLE
GH-813 remove thread locking from dependencies UI

### LINKED ISSUE #813: Dialog for installing external dependencies is using thread locking which hangs sometimes
In the dialog that coordinates setting up dependencies for the workspace, thread suspending and notification was used to synchronise UI and other threads carrying out operations.

In case of unexpected issues, this was causing zombie threads and UI windows hanging, which was causing havoc in build system in the SWTBot tests.

Synchronisation has to be carried in other way, to avoid hanging windows.

### PR DESCRIPTION
**Pull request by yoosiba**
resolves #813

### CODE REVIEW

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesInstallHelper.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+import com.google.inject.Inject;
+
+/**
+ * Helper for installing npm dependencies.
+ */
+public class ExternalLibrariesInstallHelper {
+
+	@Inject
+	private ProjectDependenciesHelper dependneciesHelper;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesInstallHelper.java]: can you also adjust the dictionary regarding `dependNecies`?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesWizard.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+		return true;
+	}
+
+	private static void showErrorMessage() {
+		UIUtils.getDisplay().asyncExec(() -> MessageDialog.openError(
+				UIUtils.getShell(),
+				"Setting up external libraries failed.",
+				"Error while setting up external libraries.\n"
+						+ "Please check your Error Log view for the detailed log about the failure.\n"
+						+ " (note that autobuild is " + (AutobuildUtils.get() ? "on" : "off") + ")"));
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesWizard.java]: multiline statements... my usual comment: using intermediate variables would improve readability and would not need more lines of code. (same below)

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+import org.eclipse.n4js.utils.StatusHelper;
+
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+import com.google.inject.Provider;
+
+/**
+ * Runnable collector of the workspace projects setting files. Note single instance single use, not thread safe.
+ */
+public class RunnableInstallDependnecies implements IRunnableWithProgress {
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java]: N

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	@Override
+	synchronized public void run(IProgressMonitor pmonitor) {
+		final SubMonitor monitor = SubMonitor.convert(pmonitor, 100);
+		final SubMonitor subMonitor2 = monitor.split(1);
+
+		final boolean wasAutoBuilding = AutobuildUtils.get();
+
+		AutobuildUtils.turnOff();
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java]: you can use the Autoclosable to temporarily turn off autobuild

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableSettingsFilesLocator.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import org.eclipse.core.runtime.SubMonitor;
+import org.eclipse.jface.operation.IRunnableWithProgress;
+
+/**
+ * Runnable collector of the workspace projects setting files.
+ */
+public class RunnableSettingsFilesLocator implements IRunnableWithProgress {
+	private static final Logger LOGGER = Logger.getLogger(RunnableSettingsFilesLocator.class);
+	// TODO seems redundant
+	ProjectsSettingsFilesLocator files = null;
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableSettingsFilesLocator.java]: make it private?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+	}
+
+	/** Scan file system for the relevant {@code .npmrc} files. */
+	private void discoverConfigFilesAndUpdate() {
+		RunnableSettingsFilesLocator runnableSettingsFilesLocator = new RunnableSettingsFilesLocator();
+		try {
+			getContainer().run(true, true, runnableSettingsFilesLocator);
+		} catch (InvocationTargetException | InterruptedException e) {
+			LOGGER.error("Error during configuration files discovery.", e);
+			UIUtils.getDisplay().asyncExec(() -> MessageDialog.openError(
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: multiline statements 💯

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+		if (items.length >= 2) {
+			String text = Strings.nullToEmpty(table.getToolTipText());
+			if (!text.isEmpty())
+				text = " :: " + text;
+			throw new RuntimeException("Multiple selections are not supported" + text);
+		}
+
+		if (items.length == 1)
+			if (items[0] != null)
+				// first is the display name, second is data
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: omitting braces... well, but two times in a row? and without an else block 😕

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmCLI.java [hunk 8]
```diff
@@ -114,8 +115,13 @@ public boolean invalidPackageVersion(String packageVersion) {
 		Collection<LibraryChange> actualChanges = new LinkedHashSet<>();
 		File installPath = new File(locationProvider.getTargetPlatformInstallLocation());
 
+		final String jobName = addressedType.name().toLowerCase();
```
Comment (mmews-n4) [inline plugins/org.eclipse.n4js/src/org/eclipse/n4js/external/NpmCLI.java]: 🥇

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java (hunk 7 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: I don't like `else` parts, prefer pattern matching or `switch` cases if the are at least like in Xtend.
I don't like braces, semicolons or other things required to reason about code when you print it in black and white.
I want language that is whitespace sensitive *and* editor that does expression+scope based colouring.


Anyway, do you feel like adding `{{}}` is necessary? I can add it.

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java (hunk 6 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: will change it

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableSettingsFilesLocator.java (hunk 5 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableSettingsFilesLocator.java]: yes

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java (hunk 4 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java]: but I want to close it *only* if there will be no exceptions... so no?

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java (hunk 3 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java]: Y

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesWizard.java (hunk 2 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesWizard.java]: ok

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesInstallHelper.java (hunk 1 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/ExternalLibrariesInstallHelper.java]: good idea

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java (hunk 7 above)
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: I meant the combination of nested if-statements _and_ having no braces makes it strange. But I see your point with the pattern above. Do as you prefer.

Comment (kduske-n4) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: or `if (items.length == 1 && items[0] != null)`

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java (hunk 4 above)
Comment (mmews-n4) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/RunnableInstallDependnecies.java]: k

File: plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java (hunk 7 above)
Comment (yoosiba) [reply plugins/org.eclipse.n4js.ui/src/org/eclipse/n4js/ui/wizard/dependencies/SetupOptionsPage.java]: then let's do it like @kduske-n4

**Pull request merged**